src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,100 +152,125 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
+            try
             {
-                return false;
-            }
+                var (loadedFile, hasAllInformation, preferredBuildHostKind, actualBuildHostKind) = await TryLoadProjectAsync(buildHostProcessManager, projectPath, cancellationToken);
+                if (preferredBuildHostKind != actualBuildHostKind)
+                    preferredBuildHostKindThatWeDidNotGet = preferredBuildHostKind;
 
-            var existingProjects = _loadedProjects.GetOrAdd(projectPath, static _ => []);
+                if (loadedFile is null)
+                {
+                    _logger.LogWarning($"Unable to load project '{projectPath}'.");
+                    return false;
+                }
 
-            Dictionary<ProjectFileInfo, ProjectLoadTelemetryReporter.TelemetryInfo> telemetryInfos = [];
-            var needsRestore = false;
+                var diagnosticLogItems = await loadedFile.GetDiagnosticLogItemsAsync(cancellationToken);
+                if (diagnosticLogItems.Any(item => item.Kind is DiagnosticLogItemKind.Error))
+                {
+                    await LogDiagnosticsAsync(diagnosticLogItems);
+                    // We have total failures in evaluation, no point in continuing.
+                    return false;
+                }
 
-            foreach (var loadedProjectInfo in loadedProjectInfos)
-            {
-                // If we already have the project with this same target framework, just update it
-                var existingProject = existingProjects.Find(p => p.GetTargetFramework() == loadedProjectInfo.TargetFramework);
-                bool targetNeedsRestore;
-                ProjectLoadTelemetryReporter.TelemetryInfo targetTelemetryInfo;
+                var loadedProjectInfos = await loadedFile.GetProjectFileInfosAsync(cancellationToken);
 
-                if (existingProject != null)
+                // The out-of-proc build host supports more languages than we may actually have Workspace binaries for, so ensure we can actually process that
+                // language in-process.
+                var projectLanguage = loadedProjectInfos.FirstOrDefault()?.Language;
+                if (projectLanguage != null && ProjectFactory.Workspace.Services.GetLanguageService<ICommandLineParserService>(projectLanguage) == null)
                 {
-                    (targetTelemetryInfo, targetNeedsRestore) = await existingProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+                    return false;
                 }
-                else
+
+                Dictionary<ProjectFileInfo, ProjectLoadTelemetryReporter.TelemetryInfo> telemetryInfos = [];
+                var needsRestore = false;
+
+                using var _ = await loadedProjectSet.Semaphore.DisposableWaitAsync(cancellationToken);
+                // last chance for someone to cancel out from under us.
```

Comment:
what do you mean by 'last chance'?  if things should not be cancellable from this point on, extract the rest into a method that explicitly states it doesn't take cancellation, that way nothing can accidentally read the wrong thing.